### PR TITLE
Adding more events to ML's core

### PR DESCRIPTION
**Pull request by slxdy**
Adding more events to ML's core would give Melon plugins more freedom and would allow the start screen to manage itself instead of relying on calls from the core.

### CODE REVIEW

**Comment by SlidyDev** [comment]
Oh ye, i forgot bout this 💀

**Comment by SlidyDev** [comment]
Bro ffs how did i miss so many errors

**Comment by SlidyDev** [comment]
I fucking broke something and now mono is crying

**Comment by SlidyDev** [comment]
Who thought setting some events up would be this much pain

**Comment by SlidyDev** [comment]
fuck you too compiler

**Comment by SlidyDev** [comment]
Wait why tf did my stupid IDE let me use a .net framework 4 reference

**Comment by SlidyDev** [comment]
SAAAAAAAAAAAAAAAAAAM

File: MelonLoader/Lemons/LemonEnumerator.cs [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
         private T[] LemonPatch;
         private int NextLemon = 0;
 
+        /// <summary>
+        /// Creates a new instance of <see cref="LemonEnumerator{T}"/> with a new copy of '<paramref name="lemons"/>'.
+        /// </summary>
         public LemonEnumerator(T[] lemons)
-            => LemonPatch = lemons.ToArray();
+        {
+            LemonPatch = new T[lemons.Length];
```
Comment (SamboyCoding) [inline MelonLoader/Lemons/LemonEnumerator.cs]: Having just dug around in the runtime reference source code, this is functionally identical to lemons.ToArray(), as ToArray has special handling for when `source` is an array itself, which ends up calling System.Array.CopyTo, which in turn calls System.Array.Copy. Basically, revert this, as it's more code for no benefit.

File: MelonLoader/MelonEvents.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        /// </summary>
+        public readonly static MelonEvent OnApplicationStart = new MelonEvent(true);
+
+        /// <summary>
+        /// Occurs when the Engine is fully initialized.
+        /// </summary>
+        public readonly static MelonEvent OnApplicationLateStart = new MelonEvent(true);
+
+        /// <summary>
+        /// Occurs when... well, the name pretty much says it.
```
Comment (SamboyCoding) [inline MelonLoader/MelonEvents.cs]: A little more professional here, please. Also, please change all "Occurs when" to "Called when"

File: MelonLoader/Melons/Events/MelonAction.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+{
+    internal class MelonAction
+    {
+        internal readonly MethodInfo method;
+        internal readonly object obj;
+        internal readonly bool unsubscribeOnFirstInvokation;
+
+        internal MelonBase Melon => obj is MelonBase melon ? melon : null;
+
+        private MelonAction(Delegate singleDel, bool unsubscribeOnFirstInvokation)
```
Comment (SamboyCoding) [inline MelonLoader/Melons/Events/MelonAction.cs]: Invokation -> Invocation everywhere in this class

File: MelonLoader/Melons/Events/MelonEvent.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+        public MelonEventBase(bool oneTimeUse = false)
+        {
+            this.oneTimeUse = oneTimeUse;
+        }
+
+        public bool CheckIfSubscribed(MethodInfo method, object obj = null)
+            => actions.Exists(x => x.method == method && (obj == null || x.obj == obj));
+
+        public void Subscribe(T action, bool unsubscribeOnFirstInvokation = false)
```
Comment (SamboyCoding) [inline MelonLoader/Melons/Events/MelonEvent.cs]: Again, invokation -> invocation

File: MelonLoader/Melons/Melon.cs [hunk 5]
```diff
@@ -20,25 +18,19 @@ public abstract class Melon<T> : MelonBase where T : Melon<T>
 
         static Melon()
         {
-            System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(typeof(T).TypeHandle); // to make sure the type initializer was triggered
+            System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(typeof(T).TypeHandle); // To make sure that the type initializer of T was triggered.
         }
 
         public sealed override string MelonTypeName => TypeName;
 
-        public sealed override bool Register()
```
Comment (SamboyCoding) [inline MelonLoader/Melons/Melon.cs:28]: This is technically an API breaking change, should consult modders to make sure nobody is using this (though I doubt it)

File: MelonLoader/Melons/MelonMod.cs [hunk 6]
```diff
@@ -1,6 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
-using System.Reflection;
+#pragma warning disable 0618
```
Comment (SamboyCoding) [inline MelonLoader/Melons/MelonMod.cs]: Please add a comment stating which warning is being disabled and why

File: MelonLoader/Melons/MelonBase.cs [hunk 7]
```diff
@@ -578,61 +592,97 @@ public virtual bool Register()
...
+            MelonPreferences.OnPreferencesLoaded.Subscribe(OnPreferencesLoaded);
+            MelonPreferences.OnPreferencesLoaded.Subscribe((x) => OnPreferencesLoaded()); // No params sig
+            MelonPreferences.OnPreferencesSaved.Subscribe(OnPreferencesSaved);
+            MelonPreferences.OnPreferencesSaved.Subscribe((x) => OnPreferencesSaved()); // No params sig
+            MelonPreferences.OnPreferencesSaved.Subscribe((x) => OnModSettingsApplied());
+        }
+
         /// <summary>
-        /// If the Name is registered, returns the registered Melon. Otherwise, returns <see langword="null"/>.
+        /// Tries to find a register Melon to matches the given Info.
```
Comment (SamboyCoding) [inline MelonLoader/Melons/MelonBase.cs]: to matches -> which matches

**Comment by slxdy** [comment]
Forgot to reimplement the start screen lol

**Review by SamboyCoding — CHANGES_REQUESTED** [review]
Looks good overall, just a couple of minor changes. I've raised a query about whether we can tidy up the event invocation.

File: MelonLoader/Lemons/LemonEnumerator.cs (hunk 1 above)
Comment (slxdy) [reply MelonLoader/Lemons/LemonEnumerator.cs]: Oh alright

File: MelonLoader/MelonEvents.cs (hunk 2 above)
Comment (slxdy) [reply MelonLoader/MelonEvents.cs]: Damn it lmao. Fine

File: MelonLoader/Melons/Events/MelonAction.cs (hunk 3 above)
Comment (slxdy) [reply MelonLoader/Melons/Events/MelonAction.cs]: love english

File: MelonLoader/Melons/Events/MelonEvent.cs (hunk 4 above)
Comment (slxdy) [reply MelonLoader/Melons/Events/MelonEvent.cs]: ly2

File: MelonLoader/Melons/Melon.cs (hunk 5 above)
Comment (slxdy) [reply MelonLoader/Melons/Melon.cs:28]: Wait, can you explain what the "API breaking change" is?

File: MelonLoader/Melons/MelonMod.cs (hunk 6 above)
Comment (slxdy) [reply MelonLoader/Melons/MelonMod.cs]: Ight

File: MelonLoader/Melons/MelonBase.cs (hunk 7 above)
Comment (slxdy) [reply MelonLoader/Melons/MelonBase.cs]: bro i actually have no damn idea how i messed that comment up this bad lmao. It must have been like 3am

File: MelonLoader/Lemons/LemonEnumerator.cs (hunk 1 above)
Comment (slxdy) [reply MelonLoader/Lemons/LemonEnumerator.cs]: https://github.com/SlidyDev/MelonLoader/commit/55272e76e91cdfa64f3548308c44728783107d55

File: MelonLoader/MelonEvents.cs (hunk 2 above)
Comment (slxdy) [reply MelonLoader/MelonEvents.cs]: Also, they all start with "Occurs" cuz they're all events, but if you really mind it that much then fine

Comment (slxdy) [reply MelonLoader/MelonEvents.cs]: https://github.com/SlidyDev/MelonLoader/commit/e25c708b3ec3d15da00b057285b4b15dae17f004

File: MelonLoader/Melons/Events/MelonAction.cs (hunk 3 above)
Comment (slxdy) [reply MelonLoader/Melons/Events/MelonAction.cs]: https://github.com/SlidyDev/MelonLoader/commit/090153abfc48d0ad44176894c6bbc89eeed815c7

File: MelonLoader/Melons/Events/MelonEvent.cs (hunk 4 above)
Comment (slxdy) [reply MelonLoader/Melons/Events/MelonEvent.cs]: https://github.com/SlidyDev/MelonLoader/commit/090153abfc48d0ad44176894c6bbc89eeed815c7

File: MelonLoader/Melons/MelonMod.cs (hunk 6 above)
Comment (slxdy) [reply MelonLoader/Melons/MelonMod.cs]: https://github.com/SlidyDev/MelonLoader/commit/f2c74f48e912f175c2d7b55d70e02d3579a57d46

File: MelonLoader/Melons/MelonBase.cs (hunk 7 above)
Comment (slxdy) [reply MelonLoader/Melons/MelonBase.cs]: https://github.com/SlidyDev/MelonLoader/commit/ed9605a47d7518b65b90fe9b6c8a3cdd7ce8ff85

File: MelonLoader/Melons/Melon.cs (hunk 5 above)
Comment (SamboyCoding) [reply MelonLoader/Melons/Melon.cs:28]: As in, you've made the method less visible, so for any mod that was calling this from outside of a melon class (which, realistically, nobody SHOULD have been), that mod is now broken.

Comment (slxdy) [reply MelonLoader/Melons/Melon.cs:28]: Wait wait what? What did i make less visible

Comment (SamboyCoding) [reply MelonLoader/Melons/Melon.cs:28]: You changed this method from public to protected internal

Comment (slxdy) [reply MelonLoader/Melons/Melon.cs:28]: Nononononono. First of all, its still public. What you're looking at is a NEW overridable method called `RegisterInternal`.
image

And even if I changed the visibility of `MelonBase::Register` rn, it hasn't even been in a full release yet so I'm still allowed to do that

Comment (SamboyCoding) [reply MelonLoader/Melons/Melon.cs:28]: Right, okay, my bad, resolving this

**Review by SamboyCoding — APPROVED** [review]
Alright, I know you still have to do start screen stuff, but I'm happy with this now.

**Comment by SlidyDev** [comment]
Baka

**Pull request merged**